File: count_routes.c
/*
 * 3797. Count Routes to Climb a Rectangular Grid
 *
 * Algorithm: DP + Prefix Sums
 * - dp0[c]: paths arriving at (r,c) via vertical move or start
 * - dp1[c]: paths arriving at (r,c) via horizontal move
 * - Transition: dp0 from row above (vertical dist ≤ d), dp1 from same row (horizontal dist ≤ d)
 * - Prefix sums for O(1) range queries
 *
 * Time: O(n * m), Space: O(m)
 * Key insight: max horizontal column delta for vertical move = isqrt(d²-1),
 *              horizontal move column delta = d (must stay on same row).
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define MOD 1000000007LL

int numberOfRoutes(char** grid, int gridSize, int d) {
    int n = gridSize;
    int m = strlen(grid[0]);

    /* max column delta for vertical (upward) move: isqrt(d*d - 1) */
    int max_up_dc = (int)sqrt((double)d * d - 1.0);
    if (max_up_dc > m - 1) max_up_dc = m - 1;

    /* Precompute availability */
    int *avail = (int *)calloc(n * m, sizeof(int));
    for (int i = 0; i < n; i++)
        for (int j = 0; j < m; j++)
            avail[i * m + j] = (grid[i][j] == '.');

    /* dp0, dp1 for current row */
    long long *dp0 = (long long *)calloc(m, sizeof(long long));
    long long *dp1 = (long long *)calloc(m, sizeof(long long));
    long long *pre = (long long *)calloc(m + 1, sizeof(long long));

    /* Row 0: base case */
    for (int c = 0; c < m; c++)
        dp0[c] = avail[c];

    /* Row 0 horizontal */
    for (int c = 0; c < m; c++)
        pre[c + 1] = (pre[c] + dp0[c]) % MOD;
    for (int c = 0; c < m; c++) {
        int lo = c - d; if (lo < 0) lo = 0;
        int hi = c + d; if (hi >= m) hi = m - 1;
        dp1[c] = ((pre[hi + 1] - pre[lo] - dp0[c]) % MOD + MOD) % MOD;
        dp1[c] = (dp1[c] * avail[c]) % MOD;
    }

    /* Process rows 1 to n-1 */
    long long *new_dp0 = (long long *)calloc(m, sizeof(long long));
    long long *new_dp1 = (long long *)calloc(m, sizeof(long long));

    for (int r = 1; r < n; r++) {
        int *av = &avail[r * m];
        int *av_prev = &avail[(r - 1) * m];

        /* Prefix sum of (dp0 + dp1) from previous row */
        for (int c = 0; c < m; c++)
            pre[c + 1] = (pre[c] + dp0[c] + dp1[c]) % MOD;

        /* Vertical transitions */
        for (int c = 0; c < m; c++) {
            if (!av[c]) { new_dp0[c] = 0; continue; }
            int lo = c - max_up_dc; if (lo < 0) lo = 0;
            int hi = c + max_up_dc; if (hi >= m) hi = m - 1;
            new_dp0[c] = (pre[hi + 1] - pre[lo] + MOD) % MOD;
        }

        /* Prefix sum of new_dp0 (vertical arrivals only) */
        for (int c = 0; c < m; c++)
            pre[c + 1] = (pre[c] + new_dp0[c]) % MOD;

        /* Horizontal transitions */
        for (int c = 0; c < m; c++) {
            if (!av[c]) { new_dp1[c] = 0; continue; }
            int lo = c - d; if (lo < 0) lo = 0;
            int hi = c + d; if (hi >= m) hi = m - 1;
            new_dp1[c] = ((pre[hi + 1] - pre[lo] - new_dp0[c]) % MOD + MOD) % MOD;
            new_dp1[c] = (new_dp1[c] * av[c]) % MOD;
        }

        /* Swap */
        long long *tmp;
        tmp = dp0; dp0 = new_dp0; new_dp0 = tmp;
        tmp = dp1; dp1 = new_dp1; new_dp1 = tmp;
    }

    /* Sum dp0 + dp1 for bottom row available cells */
    long long ans = 0;
    int *av_bottom = &avail[(n - 1) * m];
    for (int c = 0; c < m; c++)
        if (av_bottom[c])
            ans = (ans + dp0[c] + dp1[c]) % MOD;

    free(avail); free(dp0); free(dp1); free(pre);
    free(new_dp0); free(new_dp1);
    return (int)ans;
}

File: count_routes.py
import math
import time

import numpy as np


class Solution:
    def countRoutes(self, grid: list[str], d: int) -> int:
        """
        DP + 前缀和 + numpy 向量化。

        状态：dp0[c] 到达 (r,c) 最后一步向上/起点
              dp1[c] 到达 (r,c) 最后一步水平

        转移：dp0[r][c] = Σ(dp0+dp1)[r-1][c']  (欧氏距离≤d)
              dp1[r][c] = Σ dp0[r][c']            (水平距离≤d, c'≠c)

        前缀和 O(1) 范围查询，numpy 向量化消除 Python 循环开销。
        时间 O(n*m)，空间 O(m)。
        """
        MOD = 10**9 + 7
        n = len(grid)
        m = len(grid[0])

        max_up_dc = math.isqrt(d * d - 1)

        avail = np.zeros((n, m), dtype=np.int64)
        for i in range(n):
            row = grid[i]
            for j in range(m):
                if row[j] == '.':
                    avail[i, j] = 1

        idx = np.arange(m, dtype=np.int64)
        hi_up = np.minimum(m, idx + max_up_dc + 1)
        lo_up = np.maximum(0, idx - max_up_dc)
        hi_hor = np.minimum(m, idx + d + 1)
        lo_hor = np.maximum(0, idx - d)

        pre = np.zeros(m + 1, dtype=np.int64)
        dp0 = avail[0].copy()
        dp1 = np.zeros(m, dtype=np.int64)

        # Row 0 horizontal
        np.cumsum(dp0, out=pre[1:])
        dp1 = ((pre[hi_hor] - pre[lo_hor] - dp0) * avail[0]) % MOD

        for r in range(1, n):
            av = avail[r]
            np.cumsum(dp0 + dp1, out=pre[1:])
            new_dp0 = ((pre[hi_up] - pre[lo_up]) * av) % MOD
            np.cumsum(new_dp0, out=pre[1:])
            new_dp1 = ((pre[hi_hor] - pre[lo_hor] - new_dp0) * av) % MOD
            dp0, dp1 = new_dp0, new_dp1

        return int((np.sum(dp0 * avail[n-1]) + np.sum(dp1 * avail[n-1])) % MOD)


def test():
    s = Solution()

    assert s.countRoutes(["..", "#."], 1) == 2
    assert s.countRoutes(["..", "#."], 2) == 4
    assert s.countRoutes(["#"], 750) == 0
    assert s.countRoutes([".."], 1) == 4
    assert s.countRoutes(["##", "##"], 1) == 0
    assert s.countRoutes(["."], 1) == 1
    assert s.countRoutes([".", ".", "."], 1) == 1
    assert s.countRoutes(["..", ".."], 750) == 16
    assert s.countRoutes(["..", ".."], 1000) == 16
    assert s.countRoutes(["...", "...", "..."], 1) == 41
    assert s.countRoutes([".#.", ".#.", ".#."], 1) == 2
    assert s.countRoutes([".#.", ".#.", ".#."], 2) == 16
    assert s.countRoutes(["..."], 1) == 7
    assert s.countRoutes(["..."], 2) == 9

    import random
    random.seed(42)
    n, m, d_val = 750, 750, 750
    big_grid = []
    for i in range(n):
        row = ''.join('.' if random.random() > 0.1 else '#' for _ in range(m))
        big_grid.append(row)

    s.countRoutes(["..", ".."], 2)

    times = []
    for _ in range(5):
        start = time.perf_counter()
        result = s.countRoutes(big_grid, d_val)
        elapsed_ms = (time.perf_counter() - start) * 1000
        times.append(elapsed_ms)

    best = min(times)
    print(f"750x750, d=750: result={result}, best={best:.1f}ms")
    print(f"Time O(n*m), Space O(m)")
    print("All 14 tests passed!")


if __name__ == "__main__":
    test()
